src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Projects/ProjectSystemCache.cs

```diff
@@ -547,5 +548,74 @@ public void Dispose()
             Dispose(true);
             GC.SuppressFinalize(this);
         }
+
+        public bool AddProjectRestoreInfoSource(ProjectNames projectNames, object restoreInfoSource)
+        {
+            if (projectNames == null)
+            {
+                throw new ArgumentNullException(nameof(projectNames));
+            }
+
+            if (projectNames.FullName == null)
+            {
+                throw new ArgumentException(
+                    message: string.Format(CultureInfo.CurrentCulture, Strings.PropertyCannotBeNull, nameof(projectNames.FullName)),
+                    paramName: nameof(projectNames));
+            }
+
+            if (restoreInfoSource == null)
+            {
+                throw new ArgumentNullException(nameof(restoreInfoSource));
+            }
+
+            _readerWriterLock.EnterWriteLock();
```

Comment:
[ReaderWriterLockSlim](https://docs.microsoft.com/en-us/dotnet/api/system.threading.readerwriterlockslim.tryenterreadlock?view=net-5.0) has overload methods that accepts timeout while waiting for read or write lock. Should we use that API instead given that few methods in this PR wait for lock during entry and release it during exit?

Reply:
The cache is already in heavy use today and we haven't really encountered any synchronization issues. 
NuGet is the only component that owns the locks and every access is controlled.

As such I'd argue that adding timeouts would be overcomplicating things.